custom-checks/checkstyle/src/main/java/org/openhab/tools/analysis/checkstyle/readme/MarkdownVisitor.java

```diff
@@ -68,106 +56,57 @@ public MarkdownVisitor(MarkdownVisitorCallback callBack, FileText fileText) {
-        String codeLiteral = fencedCodeBlock.getLiteral();
-        validateCodeSection(codeLiteral);
+        validateCodeSectionPosition(fencedCodeBlock.getLineNumber(), fencedCodeBlock.getEndLineNumber(),
+                fencedCodeBlock);
     }
 
-    /**
-     * Processes FencedCodeBlock nodes the parser returns.
-     *
-     * @param codeBlock - multiLine String containing the literal of the code block
-     */
-    private void validateCodeSection(String codeBlock) {
-        int codeStartingLineNumber = 0;
-        int codeEndingLineNumber = 0;
+    private void validateCodeSectionPosition(int zeroBasedStartLineNumber, int zeroBasedEndLineNumber,
+            Node codeBlockText) {
 
-        // Splitting the String in case it is multiline, else the array would
-        // have only one element
-        String codeSectionLines[] = codeBlock.split("\\r?\\n");
-        if (!StringUtils.isBlank(codeBlock)) {
-            try {
-                codeStartingLineNumber = callback.findLineNumber(fileText, codeSectionLines[0], currentLinePointer);
-                codeStartingLineNumber--;
-                currentLinePointer = codeStartingLineNumber;
-                // Start from the line above the code section
-                verifyBeforeCodeSection(codeStartingLineNumber);
-                codeEndingLineNumber = codeStartingLineNumber + codeSectionLines.length;
-                codeEndingLineNumber++;
-                currentLinePointer = codeEndingLineNumber;
-                verifyAfterCodeSection(codeEndingLineNumber);
-            } catch (NoResultException e) {
-                logger.error("A code section wasn't processed properly! " + codeBlock, e);
-            }
-        } else {
-            callback.log(currentLinePointer, EMPTY_CODE_BLOCK_WARNING);
-        }
-    }
+        Node codeSection = codeBlockText.getFirstChild();
 
-    private void verifyBeforeCodeSection(int codeStartingLineNumber) {
-        if (codeStartingLineNumber == 1) {
-            callback.log(codeStartingLineNumber, EMPTY_LINE_BEFORE_CODE_MSG);
-        } else {
-            // -2 because the line before the code is occupied with code section opening- ```
-            int lineBeforeCodeSection = codeStartingLineNumber - 2;
-            boolean isLineBeforeCodeSectionEmpty = StringUtils.isBlank(fileText.get(lineBeforeCodeSection));
-            if (!isLineBeforeCodeSectionEmpty) {
-                callback.log(codeStartingLineNumber, EMPTY_LINE_BEFORE_CODE_MSG);
+        // Check if the code section is empty or blank
+        if (codeSection != null && !StringUtils.isBlank(codeSection.getChars().toString())) {
+
+            // The code block is not the first line, and the previous line is not empty
+            if (zeroBasedStartLineNumber == 0 || !StringUtils.isBlank(fileText.get(zeroBasedStartLineNumber - 1))) {
+                // log the one-based line number
+                callback.log(zeroBasedStartLineNumber + 1, EMPTY_LINE_BEFORE_CODE_MSG);
```

Comment:
I don't know whether you control the callback, but maybe you should do the adding of one there, it's too easy to forget one.

Reply:
Great idea! Thanks!